subsys/net/lib/coap/coap_client.c

```diff
@@ -190,39 +241,55 @@ static enum coap_block_size coap_client_default_block_size(void)
 }
 
 static int coap_client_init_request(struct coap_client *client,
-				    struct coap_client_request *req)
+				    struct coap_client_request *req,
+				    struct coap_client_internal_request *internal_req,
+				    bool reconstruct)
 {
 	int ret = 0;
 	int i;
 
 	memset(client->send_buf, 0, sizeof(client->send_buf));
-	ret = coap_packet_init(&client->request, client->send_buf, MAX_COAP_MSG_LEN, 1,
-			       req->confirmable ? COAP_TYPE_CON : COAP_TYPE_NON_CON,
-			       COAP_TOKEN_MAX_LEN, coap_next_token(), req->method,
-			       coap_next_id());
+
+	if (reconstruct) {
+		ret = coap_packet_init(&internal_req->request, client->send_buf, MAX_COAP_MSG_LEN,
+				       1, req->confirmable ? COAP_TYPE_CON : COAP_TYPE_NON_CON,
+				       COAP_TOKEN_MAX_LEN, internal_req->request_token, req->method,
+				       internal_req->last_id);
+	} else {
+		internal_req->last_id = coap_next_id();
+		ret = coap_packet_init(&internal_req->request, client->send_buf, MAX_COAP_MSG_LEN,
+				       1, req->confirmable ? COAP_TYPE_CON : COAP_TYPE_NON_CON,
+				       COAP_TOKEN_MAX_LEN, coap_next_token(), req->method,
```

Comment:
Shouldn't the token be actually stored in `internal_req->request_token`, just like message ID?

I also wonder, whether it made sense in case of `!reconstruct` to simply create new id/token, and have a single `coap_packet_init()` call for both cases.

Reply:
Done